caddy.go

```diff
@@ -405,13 +410,10 @@ func startWithListenerFds(cdyfile Input, inst *Instance, restartFds map[string]r
 		return err
 	}
 
-	if restartFds == nil {
-		// run startup callbacks since this is not a restart
-		for _, startupFunc := range inst.onStartup {
-			err := startupFunc()
-			if err != nil {
-				return err
-			}
+	for _, startupFunc := range inst.onStartup {
+		err := startupFunc()
+		if err != nil {
+			return err
```

Comment:
Oof. You're right. Maybe we need more specific callbacks, instead of just startup, restart, and shutdown. The older versions of caddy had a OnFirstStartup callback, but even then, log and errors didn't use it (oops).

This PR has been good, I'm re-thinking quite of few of these details now. I need to reconsider the relationship between `caddy.Instance` and `httpserver.httpContext`.

I also will look into adding more specific callbacks, so that we have:

- OnFirstStartup: When the Instance starts for the first time (not during restarts)
- OnStartup: When the Instance starts, even during restarts
- OnRestart: When the Instance begins to be restarted
- OnShutdown: When the Instance stops, even during restarts
- OnFinalShutdown: When the Instance stops for good (not a restart)

What do you think of that?

If that's good, I'll be happy to merge this PR and then make these changes.